### PR TITLE
fix: add compatibility for GraalVM 22.2.0

### LINKED ISSUE #998: IllegalAccessError when trying to build Java 17 Native image
While trying to build a simple Micronaut app with a connection to a SQL (postgres) db an **IllegalAccessError** is being thrown.
Here's the stack trace
> Fatal error: java.lang.IllegalAccessError: class com.google.cloud.sql.nativeimage.CloudSqlFeature (in unnamed module @0x3c3cd7a5) cannot access class org.graalvm.nativeimage.impl.ConfigurationCondition (in module org.graalvm.sdk) because module org.graalvm.sdk does not export org.graalvm.nativeimage.impl to unnamed module @0x3c3cd7a5
	at com.google.cloud.sql.nativeimage.CloudSqlFeature.beforeAnalysis(CloudSqlFeature.java:57)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$9(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:78)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:722)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:564)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:521)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:407)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:585)
	at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)

The error comes from the dependency **implementation("com.google.cloud.sql:postgres-socket-factory:1.7.0")**

In case it helps, this is the Dockerfile being used (which is auto-generated by Micronaut):

> FROM ghcr.io/graalvm/native-image:ol7-java17-22.2.0 AS graalvm
WORKDIR /home/app
COPY layers/libs /home/app/libs
COPY layers/classes /home/app/classes
COPY layers/resources /home/app/resources
COPY layers/application.jar /home/app/application.jar
RUN mkdir /home/app/config-dirs
COPY config-dirs/generateResourcesConfigFile /home/app/config-dirs/generateResourcesConfigFile
RUN native-image -cp /home/app/libs/*.jar:/home/app/resources:/home/app/application.jar --no-fallback -H:Name=application -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk=ALL-UNNAMED -J--add-exports=org.graalvm.nativeimage.builder/com.oracle.svm.core.configure=ALL-UNNAMED -H:ConfigurationFileDirectories=/home/app/config-dirs/generateResourcesConfigFile -H:Class=com.products.Application
FROM frolvlad/alpine-glibc:alpine-3.12
RUN apk update && apk add libstdc++
COPY --from=graalvm /home/app/application /app/application
ENTRYPOINT ["/app/application"]

Let me know if you need any other info to make this error clear.

### PR DESCRIPTION
**Pull request by shubha-rajan**
## Change Description
- Remove references to `svm` dependency
- Upgrade Native Image Maven plugin
- Change scope of drivers to `compile` for native image builds

Fixes #998 , #1027

### CODE REVIEW

File: jdbc/mysql-j-8/pom.xml [hunk 1]
```diff
@@ -164,6 +164,16 @@
         </dependency>
       </dependencies>
     </profile>
+    <profile>
+      <id>native</id>
```
Comment (enocom) [inline jdbc/mysql-j-8/pom.xml:168]: Should these deps be marked as "provided" so the user can set them?

Comment (shubha-rajan) [reply jdbc/mysql-j-8/pom.xml:168]: Marking them as `provided` was what was causing one of the errors, so here we're overriding the `provided` scope with a `compile` scope.  This was the error:
```
Failures (2):
  JUnit Vintage:JdbcPostgresIamAuthIntegrationTests:pooledConnectionTest
    MethodSource [className = 'com.google.cloud.sql.postgres.JdbcPostgresIamAuthIntegrationTests', methodName = 'pooledConnectionTest', methodParameterTypes = '']
    => org.opentest4j.MultipleFailuresError: Multiple Failures (2 failures)
	java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:postgresql:///proxy_testing
	java.lang.NullPointerException: <no message>
       org.junit.vintage.engine.execution.TestRun.getStoredResultOrSuccessful(TestRun.java:200)
       org.junit.vintage.engine.execution.RunListenerAdapter.fireExecutionFinished(RunListenerAdapter.java:248)
       org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:214)
       org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:88)
       org.junit.runner.notification.SynchronizedRunListener.testFinished(SynchronizedRunListener.java:87)
       [...]
       Suppressed: java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:postgresql:///proxy_testing
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:114)
         com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:331)
         com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:114)
         com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:108)
         com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
         [...]
       Caused by: java.sql.SQLException: No suitable driver
         java.sql.DriverManager.getDriver(DriverManager.java:298)
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:106)
         [...]
       Suppressed: java.lang.NullPointerException
         com.google.cloud.sql.postgres.JdbcPostgresIamAuthIntegrationTests.dropTableIfPresent(JdbcPostgresIamAuthIntegrationTests.java:111)
         java.lang.reflect.Method.invoke(Method.java:566)
         org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
         org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
         org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
         [...]
  JUnit Vintage:JdbcPostgresIntegrationTests:pooledConnectionTest
    MethodSource [className = 'com.google.cloud.sql.postgres.JdbcPostgresIntegrationTests', methodName = 'pooledConnectionTest', methodParameterTypes = '']
    => org.opentest4j.MultipleFailuresError: Multiple Failures (2 failures)
	java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:postgresql:///proxy_testing
	java.lang.NullPointerException: <no message>
       org.junit.vintage.engine.execution.TestRun.getStoredResultOrSuccessful(TestRun.java:200)
       org.junit.vintage.engine.execution.RunListenerAdapter.fireExecutionFinished(RunListenerAdapter.java:248)
       org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:214)
       org.junit.vintage.engine.execution.RunListenerAdapter.testFinished(RunListenerAdapter.java:88)
       org.junit.runner.notification.SynchronizedRunListener.testFinished(SynchronizedRunListener.java:87)
       [...]
       Suppressed: java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:postgresql:///proxy_testing
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:114)
         com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:331)
         com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:114)
         com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:108)
         com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
         [...]
       Caused by: java.sql.SQLException: No suitable driver
         java.sql.DriverManager.getDriver(DriverManager.java:298)
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:106)
         [...]
       Suppressed: java.lang.NullPointerException
         com.google.cloud.sql.postgres.JdbcPostgresIntegrationTests.dropTableIfPresent(JdbcPostgresIntegrationTests.java:104)
         java.lang.reflect.Method.invoke(Method.java:566)
         org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
         org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
         org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
         [...]
```

**Review by enocom — APPROVED** [review]
LGTM.

**Review by mpeddada1 — COMMENTED** [review]
@shubha-rajan the use of the `svm` dependency has been debated for some time and there were some discussions to stop releasing it. For stability in the future releases, it might be better to remove references to it. We have a couple of references here: https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/blob/15e01f41e4802ff49090cd7cd62dec7b6b2c6a57/core/src/main/java/com/google/cloud/sql/nativeimage/CloudSqlFeature.java#L20-L21.

1) Here is an example of how AutomaticFeature was replaced in gax: https://github.com/googleapis/gax-java/commit/2229bfe84563bd5172e3b23c97d1eb3e73189b7a
2) The use of the `ResourceRegistration` class can be replaced with using `resource-config.json`. For example: https://github.com/googleapis/gax-java/pull/1674/files

**Comment by shubha-rajan** [comment]
> @shubha-rajan the use of the `svm` dependency has been debated for some time and there were some discussions to stop releasing it. 
I followed the example to remove the `svm` references here. PTAL

File: core/src/main/resources/META-INF/native-image/com.google.cloud.sql/native-image/resource-config.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (mpeddada1) [inline core/src/main/resources/META-INF/native-image/com.google.cloud.sql/native-image/resource-config.json:1]: nit: how about we name this directory`META-INF/native-image/com.google.cloud.sql/jdbc-socket-factory-parent/native-image.properties`? For more info: https://www.graalvm.org/22.1/reference-manual/native-image/BuildConfiguration/#embedding-a-configuration-file

File: jdbc/mysql-j-8/pom.xml [hunk 3]
```diff
@@ -164,6 +164,16 @@
...
+      <dependencies>
+        <dependency>
+          <groupId>mysql</groupId>
+          <artifactId>mysql-connector-java</artifactId>
+          <version>8.0.17</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (shubha-rajan) [reply jdbc/mysql-j-8/pom.xml:167]: No it's a different one. "No suitable driver found." I started seeing it after upgrading the maven plugin.

[the remark this answers is not in the document]
Comment (shubha-rajan) [reply jdbc/mysql-j-8/pom.xml:167]:
```
       Suppressed: java.lang.RuntimeException: Failed to get driver instance for jdbcUrl=jdbc:postgresql:///proxy_testing
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:114)
         com.zaxxer.hikari.pool.PoolBase.initializeDataSource(PoolBase.java:331)
         com.zaxxer.hikari.pool.PoolBase.<init>(PoolBase.java:114)
         com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:108)
         com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)
         [...]
       Caused by: java.sql.SQLException: No suitable driver
         java.sql.DriverManager.getDriver(DriverManager.java:298)
         com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:106)
         [...]
       Suppressed: java.lang.NullPointerException
         com.google.cloud.sql.postgres.JdbcPostgresIamAuthIntegrationTests.dropTableIfPresent(JdbcPostgresIamAuthIntegrationTests.java:111)
         java.lang.reflect.Method.invoke(Method.java:566)
         org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
         org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
         org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
         [...]

```

File: core/src/main/resources/META-INF/native-image/com.google.cloud.sql/native-image/resource-config.json (hunk 2 above)
Comment (shubha-rajan) [reply core/src/main/resources/META-INF/native-image/com.google.cloud.sql/native-image/resource-config.json]: done

**Review by mpeddada1 — APPROVED** [review]
LGTM

File: jdbc/mysql-j-8/pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (shubha-rajan) [reply jdbc/mysql-j-8/pom.xml:167]: I think the issues are related. I'll put together a quick repro so we can track this.

[the remark this answers is not in the document]
Comment (shubha-rajan) [reply jdbc/mysql-j-8/pom.xml:167]: unfortunately we still have to upgrade because we get `UnsupportedFeatureException`s otherwise

**Pull request merged**